### DISCUSSION TITLE
Converting command line switches to plugins

### DISCUSSION THREAD

**Discussion by gitbls** [gitbls/sdm#98] [General]
# Overview
I'm in the process of removing several command switches and moving their functionality into plugins. This will improve code quality, reduce maintenance overhead, improve sdm's flexibility (plugins can be used on both customize and burn) and facilitate improved documentation, with only minimal user conversion costs.

Your feedback and questions are _greatly_ appreciated!

Here's the list of switches being removed (so far), and what plugin replaces them. Below that is a description of the plugin.

* `--cron-d`, `--cron-daily`, `--cron-hourly`, `--cron-weekly`, `--cron-monthly`, `--cron-systemd`, `--exports`, `--fstab`, `--journal`, `--sysctl`, `--systemd-config`, `--udev`, `--svc*disable`, and `--svc*enable` all replaced by plugin `system`
* `--bootconfig`, `--hdmi_group`, `--hdmi_mode`, `--hdmi_force_hotplug`, and `--hdmi_ignore_edid` all replaced by plugin `bootconfig`
* `--dtoverlay` and `--dtparam` replaced by plugin `dtstuff`
* `--bootset` replaced by plugin `raspiconfig`

## Plugin descriptions

Most (all?) arguments directly relate to their prior switch-based versions. All the plugins can be provided multiple times on the command line if desired, and can be used on both customize and burn.

### `system` plugin

The `system` plugin is a collection of system-related configuration settings. You are responsible for using correct file types (e.g., .conf, .rules, etc), the plugin does no checking/modification of file types.

#### Arguments

* **cron-d** &mdash; Comma-separated list of files to copy to /etc/cron.d
* **cron-daily** &mdash; Comma-separated list of files to copy to /etc/cron.daily
* **cron-hourly** &mdash; Comma-separated list of files to copy to /etc/cron.hourly
* **cron-weekly** &mdash; Comma-separated list of files to copy to /etc/cron.weekly
* **cron-monthly** &mdash; Comma-separated list of files to copy to /etc/cron.monthly
* **cron-systemd** &mdash; Takes no value. Switches from using cron to systemd-based cron timers
* **exports** &mdash; Comma-separated list of files to append to /etc/exports
* **fstab** &mdash; Comma-separated list of files to append to /etc/fstab
* **journal** &mdash; Configure systemd journal
* **sysctl** &mdash; Comma-separated list of files to copy to /etc/sysctl.d
* **systemd-config** &mdash; Comma-separated list of type:file, where type is one of *login*, *network*, *resolve*, *system*, *timesync*, or *user*. Copies the provided file to /etc/systemd/*type*.conf.d
* **udev** &mdash; Comma-separated list of files to copy to /etc/udev/rules.d
* **service-disable** &mdash; Comma-separated list of services to disable
* **service-enable** &mdash; Comma-separated list of services to enable

#### Examples

* `--plugin system:"cron-d=/path/to/crondscript|exports=/path/to/e1,/path/to/e2"`
* `--plugin system:"systemd-config=timesync=/path/to/timesync.conf|service-disable=svc1,svc2"`

### `bootconfig` plugin

The `bootconfig` plugin configures the contents of /boot/config.txt.

#### Arguments

* **inline** &mdash; If `inline` is provided as an argument (does not take a value), the plugin will replace existing settings in config.txt (if they exist) with any new value provided to the plugin. If it doesn't exist, or if `inline` is not provided, new arguments are appended to the end of the file.

* **section** &mdash; The `section` argument takes a value like `[pi4]`, and appends the section value to the end of config.txt.

* **somename=somevalue** &mdash; All other key/value settings are presumed to be settings in config.txt and added to it. There is no validity checking, so typos are propagated. But, on the other hand, the `bootconfig` plugin doesn't need to be updated every time a brand new setting is added to config.txt.

#### Examples

* `--plugin bootconfig:"section=[pi4]|somesetting=somevalue"`
* `--plugin bootconfig:"inline|hdmi_group=72|hdmi_force_hotplug=1|hdmi_mode=40|hdmi_ignore_edid"` &mdash; The plugin adds the correct value for `hdmi_ignore_edid` (0xa5000080)

### `dtstuff` plugin

The `dtstuff` plugin adds `dtoverlay` and `dtparam` settings to /boot/config.txt. It's very similar to `bootconfig`, except that the settings are always added to the end of /boot/config.txt. 

#### Arguments

* **dtoverlay** &mdash; Use this keyword (no value) to specify that the settings are `dtoverlay` settings. If this is not provided the settings are added as `dtparam` settings.
* **dtparam** &mdash; The inverse of `dtoverlay`. Not really used (lack of `dtoverlay` implies `dtparam`), but happily ignored so it can be used for self-documentation.
* **dtname=something** &mdash; The specified dtoverlay or dtparam is appended to /boot/config.txt

#### Examples

* `--plugin dtstuff:"dtoverlay|gpio-foo=pin=12"` &mdash; Adds a dtoverlay statement: `dtoverlay=gpio-foo=pin=12`
* `--plugin dtstuff:"sd_poll_once"` &mdash; Adds a dtparam statement: `dtparam=sd_poll_once`

### `raspiconfig` plugin

the `raspiconfig` plugin is used to modify settings supported by `raspi-config`. This is not necessarily the complete list (done quickly), and one or two of these may not be supportable. There's more work to do on this one!

#### Arguments

* **audio**
* **audioconf**
* **blanking**
* **boot_behaviour, boot_behavior**
* **boot_order**
* **boot_splash**
* **boot_wait**
* **camera**
* **composite**
* **glamor**
* **gldriver**
* **i2c**
* **leds**
* **legacy**
* **memory_split**
* **net_names**
* **onewire**
* **overclock**
* **overlayfs**
* **overscan**
* **pi4video**
* **pixdub**
* **powerled**
* **proxy**
* **rgpio**
* **serial**
* **spi**
* **xcompmgr**

#### Examples

* `--plugin raspiconfig:"net_names=1|boot_splash=1"`

**Comment by gitbls** [comment]
Also will be moving `--lxde-config` to the `graphics` plugin, and adding `--modprobe` to the `system` plugin for one-stop "system" configuration.

**Comment by tinker2much** [comment]
Among the bootconfig values, will you accept comments? So I can add documenting lines as follows:

```
# add the below minimum, to try to keep it cooler when idle
arm_freq_min=400
``` 

`--plugin bootconfig:"inline|# add the below minimum, to try to keep it cooler when idle|arm_freq_min=400"`

**Comment by tinker2much** [comment]
Could you please confirm that you would replace an existing commented key with the one I supply, whether or not I've also changed the value?  

Existing state of boot/config.txt:
```
#dtparam=i2c_arm=on
#dtparam=spi=on

```
Desired state of boot/config.txt:
```
dtparam=i2c_arm=on
dtparam=spi=off

```
Code to request this:
`--plugin bootconfig:"inline|dtparam=i2c_arm=on|dtparam=spi=off"`

**Reply by gitbls** [reply]
Yes.

**Comment by tinker2much** [comment]
Here are the --plugin bootconfig invocations for all of my pi's where I customized anything at all. I'm generating these from diffs between a vanilla boot/config.txt and a customized one. These assume that you'll let me include comment lines - I seem to have a lot of these, especially on boxes where I was testing overclocking and left lots of notes about some attempted but not currently chosen settings...

```
--plugin bootconfig:"inline|enable_uart=0"
--plugin bootconfig:"section=[pi2]|dtparam=sd_overclock=100|initial_turbo=30|core_freq=470|sdram_freq=470|enable_uart=0"
--plugin bootconfig:"inline|gpu_mem=64"
--plugin bootconfig:"inline|# enable uart for serial console|enable_uart=1"
--plugin bootconfig:"inline|dtoverlay=gpio-shutdown,gpio_pin=3"
--plugin bootconfig:"inline|dtparam=i2c_arm=on|dtparam=spi=on|enable_uart=1|dtparam=i2c_baudrate=1000000"
--plugin bootconfig:"section=[pi2]|dtparam=sd_overclock=100|initial_turbo=30|core_freq=470|gpu_mem=60"
--plugin bootconfig:"inline|# it may be slightly faster, but runs 10 degrees C hotter unde|#arm_freq=1300|#core_freq=500|#over_voltage=4|dtoverlay=gpio-shutdown,gpio_pin=5|# Run in 64-bit mode|arm_64bit=1"
--plugin bootconfig:"section=[pi3]|dtoverlay=rpi-sense |#  doesn't work; see complaint via dmesg | grep mmc|#dtoverlay=sdhost,overclock_50=100"
--plugin bootconfig:"section=[pi3+]|temp_soft_limit=70"
--plugin bootconfig:"inline|#arm_freq=1500 # in heatsink case, gcc_compile temperature mod|#arm_freq=1800|arm_freq=1900|#arm_freq=2000 # in heatsink case, gcc_compile will hit 80+ an|# add the below minimum, to try to keep it cooler when idle|arm_freq_min=400|dtoverlay=gpio-shutdown,gpio_pin=3|# Run in 64-bit mode|arm_64bit=1|# the following is for the official fan - not using that as of|# dtoverlay=gpio-fan,gpiopin=14,temp=80000"
--plugin bootconfig:"inline|#arm_freq=2200 locks up|#arm_freq=2100 fan constant under stressit, temp creeping up |#arm_freq=2000 fan keeps up but cant get below 75|#arm_freq=1900 fan keeps up, sticks at 73|# add the below minimum, to try to keep it cooler when idle|arm_freq_min=400|# Run in 64-bit mode|arm_64bit=1|dtoverlay=gpio-fan,gpiopin=14,temp=80000"
--plugin bootconfig:"inline|#arm_freq=1700|arm_freq=1900|#arm_freq=2000   # perhaps causing infrequent lockups ???|# add the below minimum, to try to keep it cooler when idle|arm_freq_min=400|dtparam=i2c_arm=off|dtoverlay=gpio-shutdown,gpio_pin=3|# Run in 64-bit mode|arm_64bit=1|dtoverlay=argonone|hdmi_enable_4kp60=1"
--plugin bootconfig:"inline|dtparam=i2c_arm=on|# Run in 64-bit mode|arm_64bit=1|gpu_mem=48"
--plugin bootconfig:"inline|arm_freq=900|gpu_mem=32"
--plugin bootconfig:"inline|dtoverlay=gpio-shutdown,gpio_pin=3"

```
The only trouble I see is where a comment line included a pipe. That's certainly on me.

**Reply by gitbls** [reply]
I'll have to look at this in more detail to comment, but here are some thoughts I had yesterday on this.

First, while I was in my thinking place (aka the shower), I concluded that a `comment=` item would be useful. I spent the next few hours finding where sdm was mis-handling backslashes in `read` commands because I wanted to have newline (\n) work.  Since everything is processed in sequential order, you can add a comment line above a config line with `comment=` (seemed useful to me, anyhow). Should be able to have comments at the end of the line after the directive, will need to validate that.

I also added `disable=` that comments out a single line (e.g., `disable=arm_64bit=1`), which led me to: since there aren't many uncommented settings in the stock config.txt, why not add a command to simply reset config.txt to an empty file and then tinker2much can add exactly what he wants? That would eliminate the need for using `inline`, give you full control over the contents, and therefore, easier to manage. Thoughts on that? (the original config.txt would be renamed and preserved, of course). You'd do something like:
```
--plugin bootconfig:"reset"
--plugin bootconfig:etc"
--plugin bootconfig:etc etc etc"
```
which would fully define the contents. Of course, instead of a string of bootconfig plugins you could also build your own config.txt and drop it into place, but what's the fun in that? 🤷‍♂️

Noticed in your example above that you've used `dtoverlay` and `dtparam` with the `bootconfig` plugin, which is great. The `dtstuff` plugin is a bit more prescriptive, and less flexible. Perhaps it's not really needed?

And yes, comments with pipes will be problematic as currently implemented.

Thanks for looking at this! I don't do a lot to my config.txt, so really appreciate your thoughts on it.

**Comment by tinker2much** [comment]
oops, I forgot that I have to to put the device tree stuff in their own plugin

**Reply by gitbls** [reply]
Actually you don't. the `dtoverlay`/`dtparam` stuff works with `bootconfig`, so as I mentioned above, maybe a separate plugin isn't needed.

**Comment by tinker2much** [comment]
Please clarify about "comment=". I was imagining something else - allowing comments by checking if the first character of a pipe-separated bootconfig segment is a "#", if so then you just slug that segment out there as its own line.

**Reply by gitbls** [reply]
That would be great, but requires a bit more work. I was trying to avoid churning up the `plugin_getargs` function, but there might be a way to apply a sharp scalpel to it without breaking the world. Comment= was an easy way to avoid that 😄 I'll have a look at it.

**Reply by gitbls** [reply]
It was a huge 5 or 6 line change. Underneath the covers it's implemented as a `comment=`, so you will see that in the keys listing.

Also, an "argument" that starts with "\n" will be processed as a `comment=` with the comment string "\n# rest of string" so a blank line can be inserted before the comment.

Thx for the nudge, I like it.

**Comment by tinker2much** [comment]
> I also added disable= that comments out a single line (e.g., disable=arm_64bit=1), which led me to: since there aren't many uncommented settings in the stock config.txt, why not add a command to simply reset config.txt to an empty file

I've never thought about an empty config.txt! Would that work? I haven't thought about why those few uncommented lines are there either... It almost doesn't make sense - if some key's got a default, why mention any value at all?

**Reply by gitbls** [reply]
Well, you don't want it empty when the system boots, because there are a couple of settings in there I'm pretty sure that you want, but you certainly don't need to start with the one that they supply, as long as yours has the important settings in it.

**Reply by tinker2much** [reply]
"the important settings" being what - just those typically uncommented in the standard config.txt? Are these defined somewhere? I HAVE recently skimmed over the very detailed official documentation on config.txt - who knew there were so many options! - but I don't remember anything about "these are required".

**Comment by gitbls** [comment]
> "the important settings" being what - just those typically uncommented in the standard config.txt? Are these defined somewhere? I HAVE recently skimmed over the very detailed official documentation on config.txt - who knew there were so many options! - but I don't remember anything about "these are required".

I was loosely thinking that "required settings" are those that are in the distributed /boot/config.txt as downloaded, before any customization:
```
dtparam=audio=on
camera_auto_detect=1
display_auto_detect=1
dtoverlay=vc4-kms-v3d
max_framebuffers=2
arm_64bit=1
disable_overscan=1

[cm4]
otg_mode=1

[pi4]
arm_boost=1
```
Clearly a couple of these are not _**absolutely required**_ but are for special requirements such as using the camera or on a cm4. Looking at the above, seems to me that the absolute minimal set of the above for a _standard_ pi would be
```
display_auto_detect=1
dtoverlay=vc4-kms-v3d
max_framebuffers=2
arm_64bit=1
disable_overscan=1
[pi4]
arm_boost=1
```
modulo removing arm_64bit on pis not capable of running 64bit.

**Comment by gitbls** [comment]
Given that there are so few "required settings", I'm now thinking that the correct approach is to NOT have the `inline` argument.

Instead, add a `reset` argument that a) saves the existing config.txt and b) creates a new empty one. An example in the documentation will show how to create config.txt with the "required settings" (as they exist in the downloaded IMGs.

This seems easier to me than the initial approach described above.

Thoughts?

**Comment by gitbls** [comment]
One last thing. For your planning, I'm going to be unable to work on this for a few days, so I'm not planning to release all this until sometime the week of 9/11. 

I'll have a much better idea on the schedule on or about 9/9.

**Comment by tinker2much** [comment]
> Given that there are so few "required settings", I'm now thinking that the correct approach is to NOT have the `inline` argument.
> 
> Instead, add a `reset` argument that a) saves the existing config.txt and b) creates a new empty one. An example in the documentation will show how to create config.txt with the "required settings" (as they exist in the downloaded IMGs.
> 
> This seems easier to me than the initial approach described above.
> 
> Thoughts?

I can say a definite "not sure". If you just want to insert or modify a single line, the originally proposed "inline" scheme would be easiest. If you wanted to insert a lot of lines, or make complex modifications, the "reset" scheme would be easiest.

**Reply by gitbls** [reply]
OK, use cases for both, so both `inline` and `reset` it is. 😄

**Comment by tinker2much** [comment]
> One last thing. For your planning, I'm going to be unable to work on this for a few days, so I'm not planning to release all this until sometime the week of 9/11.
> 
> I'll have a much better idea on the schedule on or about 9/9.

Thanks for the heads up. I appreciate everything you're doing. I can always test (or even deploy) with the previously released code. I wonder when RPT will ever release bookworm? Maybe we've got all year...   ;-)

**Reply by gitbls** [reply]
If my past record holds it will be released while I'm offline for the next few days 😵 

While there are things in Bookworm that seem better, and newer versions and features are always fun,, there are of course likely to be some new bugs for us all to find. So, while I'm looking forward to getting the new release, as always I'll cautiously upgrade my "servers". "Workstations" go immediately 😄

**Comment by gitbls** [comment]
I made the mistake of staring at the rat's nest of code that handled all the user account stuff and decided to overhaul that along with all the previously mentioned changes.

So...The following switches are ALSO being removed from sdm and functionality is moving to a new plugin: `--user`, `--uid`, `--password-pi`, `--password-root`, `--password-same`, `--password-user`, `--rootpwd` `--groups`, `--nopassword`, `--nouser`, and `--rename-pi` 

The new plugin, `user`, can be used to add or delete users, and change the password for an existing user:

`--plugin user:"adduser=newusername|password=pwd|homedir=/path/to/home|"` Other arguments include (I'm just being lazy here, they can all be specified to the plugin):
* `deluser=username` &mdash; Delete the specified user
* `Group` &mdash; Primary group for user being added
* `groups` &mdash; Other groups to be added to user
* `nochown` &mdash; Do not chown the home directory (with -R) to the newly created user
* `nohomedir` &mdash; Do not create a home directory
* `noskel` &mdash; Do not copy the /etc/skel files into the new home directory
* `nosudo` &mdash; Do not enable this user to do sudo
* `redact` &mdash; Redact passwords from /etc/sdm/cparams and /etc/sdm/history in the post-install phase
* `rootpwd` &mdash; Use this same password for the root user
* `setpassword=username` &mdash; Set the provided password for the given user
* `samba` &mdash; Enable this user for Samba (if installed)
* `smbpasswd` &mdash; SMB password for this user
* `uid` &mdash; UID for the user to be added
* `userlist` &mdash; Pointer to a file containing users to be added, one per line (in plugin argument format)
* `log=/path/to/logfile` &mdash; Log operations. Path to log provided is on the Host system, not in the IMG

Only one of `adduser`, `deluser`, and `setpassword` can be provided to a single instantiation of the `user` plugin, but the `userlist` file can contain a mixture of these 3. If more than one is specified, the first one found (in the order add/del/setpassword) is processed.

A few observant people may notice that this looks somewhat like the `addusers` plugin. Indeed, it is. The `addusers` plugin is being retired in favor of the new, updated `user` plugin, which is named consistently with the `graphics`, `network`, and `system` plugins.

There is no way to rename the `pi` user, but it's quite easy to use `--plugin user:"deluser=pi" --plugin user:"adduser=bls|...|", which has the same effect. 

Because of this change, the `$myuser` variable is slightly changed as well. The first time the `user` plugin adds a user, it sets the `$myuser` variable for use by other plugins, etc. So, best practice is to have the `user` plugin used to add the primary user be very early on the command line, before any other plugins. Except, of course in the specific case noted above (`deluser` followed by `adduser`).

I'm still putting finishing touches on the `user` plugin, so things may change slightly. If you have any questions or feedback on this, please let me know ASAP.
